// markdown/lib/markdown.js
/*=====================================================================*/
/*    .../project/hop/3.0.x/node_modules/markdown/lib/markdown.js      */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Jul  6 06:11:11 2014                          */
/*    Last change :  Tue Jul 14 07:26:18 2015 (serrano)                */
/*    Copyright   :  2014-15 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HOP markdown parser/generator                                    */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Bigloo module clause to spawn hopjs MARKDOWN initialization      */
/*---------------------------------------------------------------------*/
var hop = require( "hop" );
var path = require( "path" );
var markdown = require( "./_markdown.hop" );

/*---------------------------------------------------------------------*/
/*    XML tags                                                         */
/*---------------------------------------------------------------------*/
exports.H1 = markdown[ "MARKDOWN:H1" ];
exports.H2 = markdown[ "MARKDOWN:H2" ];
exports.H3 = markdown[ "MARKDOWN:H3" ];
exports.H4 = markdown[ "MARKDOWN:H4" ];
exports.H5 = markdown[ "MARKDOWN:H5" ];
exports.CODE = markdown[ "MARKDOWN:CODE" ];
exports.PRE = markdown[ "MARKDOWN:PRE" ];

/*---------------------------------------------------------------------*/
/*    parser ...                                                       */
/*---------------------------------------------------------------------*/
exports.parseFile = markdon[ "parse-file" ];

/*---------------------------------------------------------------------*/
/*    generators                                                       */
/*---------------------------------------------------------------------*/
exports.toMarkdown = markdon[ "xml->markdown" ];

exports.css = path.join( hop.shareDir, "hop-markdown.hss" );
exports.jscript = path.join( hop.shareDir, "hop-markdown.scm" );
