// tree/lib/tree.js
/*=====================================================================*/
/*    serrano/prgm/project/hop/3.0.x/node_modules/tree/lib/tree.js     */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Jul  6 06:11:11 2014                          */
/*    Last change :  Mon Jul  6 13:11:13 2015 (serrano)                */
/*    Copyright   :  2014-15 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HOP tree HTML widget                                             */
/*=====================================================================*/

var hop = require( "hop" );
var tree = require( "./_tree.hop" );
var path = require( "path" );

exports.TREE = tree.TREE;
exports.TRHEAD = tree.TRHEAD;
exports.TRBODY = tree.TRBODY;
exports.TRLEAF = tree.TRLEAF;

exports.close = tree.close;
exports.open = tree.open;
exports.toggle = tree.toggle;
exports.reset = tree.reset;
exports.selection = tree.selection;
exports.selectNext = tree.selectNext;
exports.selectPrev = tree.selectPrev;
exports.selectAll = tree.selectAll;

exports.css = path.join( hop.shareDir, "hop-tree.css" );
exports.jscript = path.join( hop.shareDir, "hop-tree.js" );
