// hopc/node/hopc.mjs
/*=====================================================================*/
/*    serrano/prgm/project/hop/hop/node_modules/hopc/node/hopc.mjs     */
/*    -------------------------------------------------------------    */
/*    Author      :  manuel serrano                                    */
/*    Creation    :  Tue Oct 24 09:50:25 2023                          */
/*    Last change :  Wed May 22 16:17:55 2024 (serrano)                */
/*    Copyright   :  2023-24 manuel serrano                            */
/*    -------------------------------------------------------------    */
/*    hopc compiler                                                    */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Import/Export                                                    */
/*---------------------------------------------------------------------*/
import * as path from "path";
import * as fs from "fs";
import { compile as jsCompile } from "./hopc-js-compiler.mjs";

export * from "./ast.mjs";
export * from "./list.mjs";
export { locToLnumCol, $ioParseError } from "./error.mjs";
export { Parser, compileErrorHandler } from "./hopc-js-compiler.mjs";
export { generate } from "./generate.mjs";

/*---------------------------------------------------------------------*/
/*    usage ...                                                        */
/*---------------------------------------------------------------------*/
function usage(argv, i) {
   if (i >= 0 && argv[i] !== "--help") {
      console.log(`hopc: wrong option "${argv[i]}" (see hopc --help)`);
   } else {
      console.log("hopc usage: source [-o target] [--help]");
   }
   process.exit(0);
}

/*---------------------------------------------------------------------*/
/*    hopCompile ...                                                   */
/*---------------------------------------------------------------------*/
async function hopCompile(src, tgt, conf) {
   let tmp = undefined;
   
   if (!tgt) {
      tmp = src.replace(/\.hop\./, ".");
   } else if (tgt === "-") {
      tmp = path.join(path.dirname(src), "._" + path.basename(src));
   } else {
      tmp = tgt;
   }

   const jsfile = await jsCompile(src, tmp, conf);

   if (!tgt) {
      console.log(fs.readFileSync(jsfile).toString());
      fs.unlinkSync(jsfile);
   }
}

/*---------------------------------------------------------------------*/
/*    relativeSourceMap ...                                            */
/*---------------------------------------------------------------------*/
function relativeSourceMap(filename, dirname) {
   const smap = JSON.parse(fs.readFileSync(filename));
   smap.sources = smap.sources.map(f=> dirname + "/" + f);
   fs.writeFileSync(filename, JSON.stringify(smap));
}
   
/*---------------------------------------------------------------------*/
/*    main ...                                                         */
/*---------------------------------------------------------------------*/
async function main(argv) {
   let src = undefined;
   let tgt = undefined;
   let i = 2;
   let conf = { typescript: false, importFrom: "@hop/hop" };

   while (i < argv.length) {
      switch (argv[i]) {
         case "--help": {
            usage(argv, -1);
         }
         case "-o": {
            if (i < argv.length - 1) {
               tgt = argv[i+1];
               i += 2;
               break;
            } else {
               usage(argv, i);
            }
         }
	 case "--client": {
	    conf.importFrom = "@hop/hop/client.mjs";
	    i++;
	    break;
	 }
         default: {
            if (src) {
               usage(argv, i);
            } else {
               src = argv[i];
               i++;
               break;
            }
         }
      }
   }

   if (!src) {
      usage(argv, -1);
   } else {
      if (src.match(/\.hop\.ts$/)) {
	 // hop-to-typescript compilation
	 conf.typescript = true;
	 await hopCompile(src, tgt, conf);
      } else if (src.match(/\.hop\.m?js$/)) {
	 // hop-to-javascript compilation
	 await hopCompile(src, tgt, conf);
      } else if (src.match(/\.ts$/)) {
	 // typescript-to-javascript compilation
	 // lazy load of the tsc wrapper
	 const tsc = await import("./ts-js-compiler.mjs");
	 const tsfile = await tsc.compile(src);

	 if (!tgt) {
	    console.log(fs.readFileSync(tsfile).toString());
	    fs.unlinkSync(tsfile);
	 } else if (tsfile !== tgt) {
	    fs.renameSync(tsfile, tgt);

	    const tsmap = tsfile + ".map";
	    if (fs.existsSync(tsmap)) {
	       const tgtmap = tgt + ".map";
	       fs.renameSync(tsmap, tgtmap);

	       const reldir =
		  path.relative(path.dirname(tgt), path.dirname(tsfile));
	       if (reldir.length !== 0) {
		  relativeSourceMap(tgtmap, reldir);
	       }
	    }
	 }
      } else {
	 throw `hopc: Don't know what do to with ${src}`;
      }
   } 
}

/*---------------------------------------------------------------------*/
/*    toplevel                                                         */
/*---------------------------------------------------------------------*/
main(process.argv)
   .catch(e => {
      console.error(e);
      process.exit(1);
   });
