// hopc/node/token.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/3.7.x/node_modules/hopc/node/token.mjs      */
/*    -------------------------------------------------------------    */
/*    Author      :  manuel serrano                                    */
/*    Creation    :  Sat Sep 16 12:31:35 2023                          */
/*    Last change :  Thu Nov 16 12:51:18 2023 (serrano)                */
/*    Copyright   :  2023 manuel serrano                               */
/*    -------------------------------------------------------------    */
/*    J2S tokens                                                       */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Token ...                                                        */
/*---------------------------------------------------------------------*/
/* @sealed */
export class Token {
   type;
   value;
   loc;

   constructor(type, value, loc) {
      this.type = type;
      this.value = value;
      this.loc = loc;
   }
}

export function makeToken(type, val, loc) {
   return new Token(type, val, loc);
}

export function tokenp(tok) {
   return tok instanceof Token;
}

export function tokenType(tok) {
   return tok.type;
}

export function tokenValue(tok) {
   return tok.value;
}

export function tokenLoc(tok) {
   return tok.loc;
}

export function queueInit(...tok) {
   return tok;
}

export function queueEmptyp(queue) {
   return queue.length === 0;
}

export function queuePush(val, queue) {
   queue.push(val);
   return queue;
}

export function queuePop(queue) {
   queue.pop();
   return queue;
}

export function queuePeek(queue) {
   return queue[queue.length - 1];
}
