// hopc/node/ts-compiler.mjs
/*=====================================================================*/
/*    .../project/hop/hop/node_modules/hopc/node/ts-compiler.mjs       */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Tue Feb 13 09:30:38 2024                          */
/*    Last change :  Mon Feb 19 10:24:24 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    Simple typescript compiler using the TSC API                     */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    module                                                           */
/*---------------------------------------------------------------------*/
import { default as ts } from "typescript";
import * as path from "node:path";
import { compileErrorLineCol, compileErrorPoint } from "./error.mjs";
import { existsSync, readFileSync, writeSync } from "node:fs";
import { load as loadSourceMap } from "./sourcemap.mjs";

export { compile };

/*---------------------------------------------------------------------*/
/*    global options                                                   */
/*---------------------------------------------------------------------*/
const HOPMODULESDIR = import.meta.url.replace(/^file:\/\//, "");

const options = {
   pretty: false,
   allowJS: true,
   target: ts.ScriptTarget.ES2020,
   module: ts.ModuleKind.ES2020,
   autorequire: true,
   moduleResolution: ts.ModuleResolutionKind.NodeJs,
   pretty: false,
   experimentalDecorators: true,
   allowSyntheticDefaultImports: true,
   sourceMap: true
};

/*---------------------------------------------------------------------*/
/*    sourceMap ...                                                    */
/*---------------------------------------------------------------------*/
function sourceMap(filename, line, col) {
   const buf = readFileSync(filename).toString();
   const m = buf.match(/^\/\/# sourceMappingURL=([^\n]+)$/m);

   if (m && existsSync(m[1])) {
      const lines = loadSourceMap(m[1]);
      if (line > 0 && line < lines.length) {
	 const seg = lines[line].find(s => s.tgtCol >= col);
	 if (seg) {
	    return { file: seg.filename, lnum: seg.srcLine, col: seg.srcCol };
	 }
      }
   }
   
   return false;
}

/*---------------------------------------------------------------------*/
/*    compile ...                                                      */
/*---------------------------------------------------------------------*/
async function compile(source) {
   try {
      const program = ts.createProgram([source], options);
      const emitResult = program.emit();
      const allDiagnostics = ts
	 .getPreEmitDiagnostics(program)
	 .concat(emitResult.diagnostics);

      allDiagnostics.forEach(diagnostic => {
	 let msg =
	    ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
	 if (diagnostic.file) {
	    let { line, character } =
	       ts.getLineAndCharacterOfPosition(diagnostic.file, diagnostic.start);

	    const { file, lnum, col } = sourceMap(source, line, character);

	    if (file) {
	       compileErrorLineCol(file, lnum, col, msg);
	    } else {
	       
	       compileErrorPoint(diagnostic.file.fileName, line + 1, col + 1, msg);
	    }
	    setTimeout(() => process.exit(1));
	 } else {
	    throw new TypeError(msg);
	 }
      });
      
      if (emitResult.emitSkipped || allDiagnostics.length > 0) {
	 process.exit(1);
	 return false;
      } else {
	 return source.replace(/\.ts$/, ".js");
      }
   } catch(e) {
      console.error("e=", e);
      process.exit(1);
      return false;
   }
}
