// hopc/node/ts-compiler.mjs
/*=====================================================================*/
/*    .../project/hop/hop/node_modules/hopc/node/ts-compiler.mjs       */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Tue Feb 13 09:30:38 2024                          */
/*    Last change :  Thu Feb 22 09:51:37 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    Simple typescript compiler using the TSC API                     */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    module                                                           */
/*---------------------------------------------------------------------*/
import { default as ts } from "typescript";
import * as path from "node:path";
import { compileErrorLineCol, compileErrorPoint } from "./error.mjs";
import { existsSync, readFileSync, writeFileSync, writeSync } from "node:fs";
import { load as loadSourceMap, encode } from "./sourcemap.mjs";

export { compile };

/*---------------------------------------------------------------------*/
/*    global options                                                   */
/*---------------------------------------------------------------------*/
const HOPMODULESDIR = import.meta.url.replace(/^file:\/\//, "");

const options = {
   pretty: false,
   allowJS: true,
   target: ts.ScriptTarget.ES2020,
   module: ts.ModuleKind.ES2020,
   autorequire: true,
   moduleResolution: ts.ModuleResolutionKind.NodeJs,
   experimentalDecorators: true,
   allowSyntheticDefaultImports: true,
   sourceMap: true
};

/*---------------------------------------------------------------------*/
/*    sourceMap ...                                                    */
/*---------------------------------------------------------------------*/
function sourceMap(filename, line, col) {
   const buf = readFileSync(filename).toString();
   const m = buf.match(/^\/\/# sourceMappingURL=([^\n]+)$/m);

   if (m && existsSync(m[1])) {
      const { sources, lines } = loadSourceMap(m[1]);
      if (line >= 0 && line < lines.length) {
	 const seg = lines[line].find(s => s.tgtCol >= col);
	 if (seg) {
	    return { file: sources[seg.srcIndex], lnum: seg.srcLine, col: seg.srcCol };
	 }
      }
   }
   
   return { file: false, lnum: -1, col: -1 };
}

/*---------------------------------------------------------------------*/
/*    remap ...                                                        */
/*---------------------------------------------------------------------*/
function remap(jsline, tslines) {
   return jsline.map(s => {
      if (s.srcLine < tslines.length) {
	 const oline = tslines[s.srcLine];
	 const fs = oline.find(t => t.tgtCol >= s.srcCol);
	 if (fs) {
	    s.srcLine = fs.srcLine;
	    s.srcCol = fs.srcCol;
	 }
      }
   });
}

/*---------------------------------------------------------------------*/
/*    reSourceMap ...                                                  */
/*    -------------------------------------------------------------    */
/*    TypeScript has generated a source map file that maps the         */
/*    .js to the .ts file. This function "remaps" to the orginal .ts   */
/*    file.                                                            */
/*---------------------------------------------------------------------*/
function reSourceMap(target, filename) {
   const buf = readFileSync(filename).toString();
   const m = buf.match(/^\/\/# sourceMappingURL=([^\n]+)$/m);

   if (m && existsSync(m[1])) {
      const jsmap = loadSourceMap(target);
      const tsmap = loadSourceMap(m[1]);
      const jslines = jsmap.lines;
      const tslines = tsmap.lines;
      jslines.forEach(l => remap(l, tslines));
      jsmap.sources = tsmap.sources;

      writeFileSync(target, JSON.stringify(encode(jsmap)));
   }
}

/*---------------------------------------------------------------------*/
/*    compile ...                                                      */
/*---------------------------------------------------------------------*/
async function compile(source) {
   try {
      const program = ts.createProgram([source], options);
      const emitResult = program.emit();
      const allDiagnostics = ts
	 .getPreEmitDiagnostics(program)
	 .concat(emitResult.diagnostics);

      allDiagnostics.forEach(diagnostic => {
	 let msg =
	    ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
	 if (diagnostic.file) {
	    let { line, character: char } =
	       ts.getLineAndCharacterOfPosition(diagnostic.file, diagnostic.start);
	    const filename = diagnostic.file.fileName;

	    if (filename === source) {
	       const { file, lnum, col } = sourceMap(source, line, char);

	       console.error("LINE=", line, "CHAR=", char, "lnum=", lnum, "col=", col);
	       if (file) {
		  compileErrorLineCol(file, lnum, col, msg);
	       } else {
		  compileErrorLineCol(filename, line, char, msg);
	       }
	    } else {
	       compileErrorLineCol(filename, line, char, msg);
	    }
	    setTimeout(() => process.exit(1));
	 } else {
	    throw new TypeError(msg);
	 }
      });
      
      if (emitResult.emitSkipped || allDiagnostics.length > 0) {
	 process.exit(1);
	 return false;
      } else {
	 const target = source.replace(/\.ts$/, ".js");
	 const targetsm = target + ".map";

	 if (existsSync(targetsm)) {
	    reSourceMap(targetsm, source);
	 }
	 
	 return target;
      }
   } catch(e) {
      console.error("e=", e);
      process.exit(1);
      return false;
   }
}
