// hopc/node/utils.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/3.7.x/node_modules/hopc/node/utils.mjs      */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Thu Dec  7 09:10:52 2023                          */
/*    Last change :  Fri Jan 19 08:16:12 2024 (serrano)                */
/*    Copyright   :  2023-24 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    Utility functions for the generated JavaScript parser            */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    The module                                                       */
/*---------------------------------------------------------------------*/
import { readFileSync } from "node:fs";

/*---------------------------------------------------------------------*/
/*    symbolp ...                                                      */
/*---------------------------------------------------------------------*/
export function symbolp(o) {
   return typeof(o) === "string";
}

export function symboltostring(x) {
   return x;
}

export function symboltostring$(x) {
   return x;
}

export function fileSeparator() {
   return "/";
}

export function makeFileName(base, file) {
   return base + "/" + file;
}

export function pwd() {
   return process.cwd();
}

export function fileNameCanonicalize$(path) {
   return path;
}

export function openInputFile(path) {
   return { filename: path, buffer: readFileSync(path).toString(), offset: 0 };
}

export function openInputString(val) {
   return { filename: false, buffer: val.toString(), offset: 0 };
}

export function inputPortName(inputPort) {
   return inputPort.filename;
}

export function inputPortPosition(inputPort) {
   return inputPort.offset;
}

export function format(fmt, ...args) {
   let fix = fmt.split("~");
   let res = fix[0];

   args.forEach((a, i) => {
      res += args[i].toString() + fix[i+1].substring(1);
   })

   return res;
}

/*---------------------------------------------------------------------*/
/*    memqArray ...                                                    */
/*---------------------------------------------------------------------*/
export function memqArray(obj, arr) {
   return arr && arr.indexOf(obj) >= 0;
}

/*---------------------------------------------------------------------*/
/*    substring ...                                                    */
/*---------------------------------------------------------------------*/
export function substring(s, offset, len) {
   return s.substring(offset, len);
}

/*---------------------------------------------------------------------*/
/*    stringPrefixp ...                                                */
/*---------------------------------------------------------------------*/
export function stringPrefixp(p, s) {
   return s.indexOf(p) === 0;
}

/*---------------------------------------------------------------------*/
/*    stringtosymbol ...                                               */
/*---------------------------------------------------------------------*/
export function stringtosymbol(s) {
   return s;
}

/*---------------------------------------------------------------------*/
/*    stringIndexRight ...                                             */
/*---------------------------------------------------------------------*/
export function stringIndexRight(s, char) {
   const i = s.lastIndexOf(char);
   if (i >= 0) {
      return i;
   } else {
      return false;
   }
}

/*---------------------------------------------------------------------*/
/*    stringUpcase ...                                                 */
/*---------------------------------------------------------------------*/
export function stringUpcase(s) {
   return s.toUpperCase();
}   

/*---------------------------------------------------------------------*/
/*    flonump ...                                                      */
/*---------------------------------------------------------------------*/
export function flonump(val) {
   return false;
}

/*---------------------------------------------------------------------*/
/*    bignump ...                                                      */
/*---------------------------------------------------------------------*/
export function bignump(val) {
   return false;
}

/*---------------------------------------------------------------------*/
/*    fixnumtollong ...                                                */
/*---------------------------------------------------------------------*/
export function fixnumtollong(val) {
   return val;
}

/*---------------------------------------------------------------------*/
/*    declUsageHasp ...                                                */
/*---------------------------------------------------------------------*/
export function declUsageHasp(decl, keys) {
   return keys.every(k => k in decl.usage);
}

/*---------------------------------------------------------------------*/
/*    gensymCnt ...                                                    */
/*---------------------------------------------------------------------*/
let gensymCnt = 0;

/*---------------------------------------------------------------------*/
/*    gensym ...                                                       */
/*---------------------------------------------------------------------*/
export function gensym() {
   return "g" + gensymCnt++;
}
