// exif/node/exif-module.js
/*=====================================================================*/
/*    .../project/hop/hop/node_modules/exif/node/exif-module.js        */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Tue Mar 26 07:54:31 2024                          */
/*    Last change :  Sun Apr  7 18:19:40 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    Exif parser (generated from bigloo:api/multimedia/Llib/exif.scm) */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    The module                                                       */
/*---------------------------------------------------------------------*/
import * as fs from "node:fs";
import { Mmap } from "./mmap.mjs";

/*---------------------------------------------------------------------*/
/*    errors                                                           */
/*---------------------------------------------------------------------*/
const $errnoIoFileNotFoundError = Error.ENOENT;
const $errnoIoParseError = -1;

/*---------------------------------------------------------------------*/
/*    error_errno ...                                                  */
/*---------------------------------------------------------------------*/
function error_errno(errno, proc, msg, obj) {
   const err = new Error(`${proc}: ${msg} -- ${obj}`);
   err.errno = errno;
   throw err;
}

/*---------------------------------------------------------------------*/
/*    $ioParseError ...                                                */
/*---------------------------------------------------------------------*/
function $ioParseError({proc, msg, obj}) {
   throw new Error(`${proc}: ${msg} -- ${obj}`);
}

/*---------------------------------------------------------------------*/
/*    warning ...                                                      */
/*---------------------------------------------------------------------*/
function warning(...vals) {
   console.log.apply(console, vals);
}

/*---------------------------------------------------------------------*/
/*    makeDate ...                                                     */
/*---------------------------------------------------------------------*/
function makeDate({sec, min, hour, day, month, year}) {
   return new Date(year, month - 1, day, hour, min, sec);
}

/*---------------------------------------------------------------------*/
/*    cons ...                                                         */
/*---------------------------------------------------------------------*/
function cons(a, d) {
   return [a, d];
}

/*---------------------------------------------------------------------*/
/*    pairp ...                                                        */
/*---------------------------------------------------------------------*/
function pairp(o) {
   Array.isArray(o) && o.length === 2;
}

/*---------------------------------------------------------------------*/
/*    car ...                                                          */
/*---------------------------------------------------------------------*/
function car(o) {
   return o[0];
}

/*---------------------------------------------------------------------*/
/*    cdr ...                                                          */
/*---------------------------------------------------------------------*/
function cdr(o) {
   return o[1];
}

/*---------------------------------------------------------------------*/
/*    stringShrink$ ...                                                */
/*---------------------------------------------------------------------*/
function stringShrink$(s, len) {
   return s.substring(0, len);
}

/*---------------------------------------------------------------------*/
/*    Exif ...                                                         */
/*---------------------------------------------------------------------*/
export function Exif(file) {
   return jpegExif(file);
}

Exif.prototype = exif.prototype;
