// filechooser/lib/filechooser.js
/*=====================================================================*/
/*    .../hop/3.0.x/node_modules/filechooser/lib/filechooser.js        */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Jul  6 06:11:11 2014                          */
/*    Last change :  Tue Sep 15 08:01:39 2015 (serrano)                */
/*    Copyright   :  2014-21 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HOP filechooser HTML widget                                      */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Bigloo module clause to spawn hopjs FILECHOOSER initialization   */
/*---------------------------------------------------------------------*/
var hop = require( "hop" );
var config = require( hop.config );
var path = require( "path" );
var filechooser = require( "./_filechooser.hop" );

const FILECHOOSER = filechooser.FILECHOOSER;

const css = path.join( config.shareDir, "hop-file.hss" );
const script = path.join( config.shareDir, "hopjs-file.js" );

/*---------------------------------------------------------------------*/
/*    nodejs module                                                    */
/*---------------------------------------------------------------------*/
exports.FILECHOOSER = FILECHOOSER;
exports.SPTAB = SPTAB;
exports.SPHEAD = SPHEAD;
exports.SPTABHEAD = SPTABHEAD;

exports.css = css;
exports.script = script;
exports.jscript = exports.script;

/*---------------------------------------------------------------------*/
/*    es6 module                                                       */
/*---------------------------------------------------------------------*/
export { FILECHOOSER, SPTAB, SPHEAD, SPTABHEAD, css, script, script as jscript };
