// config/lib/config.js
/*=====================================================================*/
/*    .../prgm/project/hop/3.0.x/node_modules/config/lib/config.js     */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Fri Jun 19 13:19:57 2015                          */
/*    Last change :  Fri Jun 26 20:27:12 2015 (serrano)                */
/*    Copyright   :  2015 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    HOP config manager                                               */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Bigloo module clause to spawn hopjs Config initialization        */
/*---------------------------------------------------------------------*/
var hop = require( "hop" );
var path = require( "path" );
var config = require( "./_config.hop" );

exports.HTTPSPrivateKey = config.HTTPSPrivateKey;
exports.HTTPSCertificate = config.HTTPSCertificate;
exports.HTTPAuthentication = config.HTTPAuthentication;

exports.xmlBackend = config.xmlBackend;

exports.maxThreads = config.maxThreads;

exports.enableKeepAlive = config.enableKeepAlive;

exports.enableWebdav = config.enableWebdav;

exports.path = config.path;
