// syslog/node/syslog.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/hop/node_modules/syslog/node/syslog.mjs     */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Wed May 29 08:23:43 2024                          */
/*    Last change :  Tue Aug 26 14:22:20 2025 (serrano)                */
/*    Copyright   :  2025 Stephen Vickers, Manuel Serrano              */
/*    -------------------------------------------------------------    */
/*    Dummy Syslog binding for Node                                    */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Import/Export                                                    */
/*---------------------------------------------------------------------*/
export { open, log, close };
export { LOG_CONS, LOG_NDELAY, LOG_ODELAY, LOG_PID };
export { LOG_AUTH, LOG_AUTHPRIV, LOG_CRON, LOG_DAEMON,
       	 LOG_FTP, LOG_KERN, 
	 LOG_LOCAL0, LOG_LOCAL1, LOG_LOCAL2, LOG_LOCAL3, LOG_LOCAL4, 
	 LOG_LOCAL5, LOG_LOCAL6, LOG_LOCAL7,
	 LOG_LPR, LOG_MAIL, LOG_NEWS, LOG_SYSLOG, LOG_USER, LOG_UUCP };
export { LOG_EMERG, LOG_ALERT, LOG_CRIT, LOG_ERR,
       	 LOG_WARNING, LOG_NOTICE, LOG_INFO, LOG_DEBUG };

/*---------------------------------------------------------------------*/
/*    constants                                                        */
/*---------------------------------------------------------------------*/
const LOG_CONS = 0, LOG_NDELAY = 1, LOG_ODELAY = 2, LOG_PID = 3;
const LOG_AUTH = 4, LOG_AUTHPRIV = 5, LOG_CRON = 6, LOG_DAEMON = 7,
  LOG_FTP = 8, LOG_KERN = 9, 
  LOG_LOCAL0 = 10, LOG_LOCAL1 = 11, LOG_LOCAL2 = 12, LOG_LOCAL3 = 13,
  LOG_LOCAL4 = 14, LOG_LOCAL5 = 15, LOG_LOCAL6 = 16, LOG_LOCAL7 = 17;
const LOG_LPR = 18, LOG_MAIL = 19, LOG_NEWS = 20, LOG_SYSLOG = 21, LOG_USER = 22, LOG_UUCP = 23;
const LOG_EMERG = 24, LOG_ALERT = 25, LOG_CRIT = 26, LOG_ERR = 27,
  LOG_WARNING = 28, LOG_NOTICE = 29, LOG_INFO = 30, LOG_DEBUG = 31;

/*---------------------------------------------------------------------*/
/*    open ...                                                         */
/*---------------------------------------------------------------------*/
function open(name, option, facility) {
}

/*---------------------------------------------------------------------*/
/*    syslog ...                                                       */
/*---------------------------------------------------------------------*/
function syslog(level, msg) {
}

/*---------------------------------------------------------------------*/
/*    close ...                                                        */
/*---------------------------------------------------------------------*/
function close() {
}
