// hop/node/hc-compiler.mjs
/*=====================================================================*/
/*    .../project/hop/3.7.x/node_modules/hop/node/hc-compiler.mjs      */
/*    -------------------------------------------------------------    */
/*    Author      :  manuel serrano                                    */
/*    Creation    :  Wed Oct 25 08:13:33 2023                          */
/*    Last change :  Wed Jan 17 13:53:18 2024 (serrano)                */
/*    Copyright   :  2023-24 manuel serrano                            */
/*    -------------------------------------------------------------    */
/*    HipHop compiler                                                  */
/*=====================================================================*/
"use strict"
"use hopscript"

/*---------------------------------------------------------------------*/
/*    es6 module                                                       */
/*---------------------------------------------------------------------*/
import { Parser, generate, $ioParseError } from "@hop/hopc";
import { writeFile } from 'node:fs/promises';
import { dirname } from "path";

/*---------------------------------------------------------------------*/
/*    language compiler (pre-processor)                                */
/*---------------------------------------------------------------------*/
const parser = new Parser();

export const parse = parser.parse.bind(parser);
export const parseString = parser.parseString.bind(parser);

/*---------------------------------------------------------------------*/
/*    HopProgram ...                                                   */
/*---------------------------------------------------------------------*/
class HopProgram {
   #source = false;
   #fragment = undefined;
   
   constructor(source, fragment) {
      this.#source = source;
      this.#fragment = fragment;
   }

   output(target) {
      const prgm = this.#fragment.toString();

      if (!target) {
	 return new Promise((res, rej) => { console.log(prgm); res(true) });
      } else {
	 const tag = `\n//# sourceMappingURL=${target}.map\n`;

	 return writeFile(target, prgm + tag);
      }
   }

   sourcemap(target) {
      const sm = {
	 version: 3,
	 file: target,
	 sources: [ this.#source ],
	 names: [],
	 mappings: this.#fragment.mappings(this.#source, target)
      }
      return writeFile(target + ".map", JSON.stringify(sm));
   }
}

/*---------------------------------------------------------------------*/
/*    compileHandler ...                                               */
/*---------------------------------------------------------------------*/
function compileHandler(err) {
   if (err && err.obj && err.obj.loc?.filename) {
      $ioParseError({
	 "fname": err.obj.loc.filename,
	 "location": err.obj.loc?.offset ?? -1,
	 "msg": "Parse error",
	 "message": err.msg || err.message,
	 forceexit: true
      });
      throw "syntax error.";
   } else if (err && err?.fname && err?.location) {
      $ioParseError({
	 "fname": err.fname,
	 "location": err.location,
	 "message": err.msg || err.message,
	 "message": err.msg,
	 forceexit: true
      });
      throw "syntax error.";
   } else if (err && err?.location) {
      $ioParseError({
	 "fname": err.location.filename,
	 "location": err.location.pos,
	 "msg": "Parse error",
	 "message": err.msg || err.message,
	 forceexit: true
      });
   } else {
      throw err;
   }
}
    
/*---------------------------------------------------------------------*/
/*    compile ...                                                      */
/*---------------------------------------------------------------------*/
export function compile(source) {
   try {
      const prog = parse(source, {}, {});
      return new HopProgram(source, generate(prog));
   } catch(err) {
      compileHandler(err);
   }
}
