// hop/node/hop-client.mjs
/*=====================================================================*/
/*    .../project/hop/hop/node_modules/hop/node/hop-client.mjs         */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Wed Jan 31 08:34:57 2024                          */
/*    Last change :  Mon Mar 25 08:22:34 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    Hop client-side part.                                            */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Module declaration                                               */
/*---------------------------------------------------------------------*/
export { server };

/*---------------------------------------------------------------------*/
/*    Frame ...                                                        */
/*---------------------------------------------------------------------*/
class Frame {
   #svc;
   #path;
   #args;
   #options;
   #header;
   
   constructor(svc, path, args, options, header) {
      this.#svc = svc;
      this.#path = path;
      this.#args = Array.from(args);
      this.#options = options;
      this.#header = header;
   }

   toString() {
      if (this.#args) {
	 let json = JSON.stringify(this.#args);
	 return this.#path + "?Hop-Serialize=json-url&Hop-Arguments="
	    + encodeURIComponent(json);
      } else {
	 return this.#path;
      }
   }

   post(opts) {
      return new Promise((res, rej) => {
	 const req = Frame.XMLHttpRequest(res, rej);

	 if (this.#args.length > 0) {
	    req.open("POST", this.#path, false);
	    req.setRequestHeader("Hop-Serialize", "json");
	    req.setRequestHeader("Hop-Arguments", JSON.stringify(this.#args));
	 } else {
	    req.open("PUT", this.#path, false);
	 }
	 
	 try {
	    req.send();
	 } catch (e) {
	    rej(e);
	 }
      });
   }

   static XMLHttpRequest(res, rej) {
      const req = new XMLHttpRequest();

      req.onreadystatechange = req.onload = function() {
	 if (this.readyState === 4) {
	    switch (this.status) {
	       case 200:
		  res(decode(req, this));
		  break;
		  
	       case 400:
	       case 407:
	       case 500:
		  rej(this.response);
		  break;
		  
	       default: {
		  if (this.status > 200 && this.status < 300) {
		     res(this.response);
		  } else {
		     rej(this.response);
		  }
		  break;
	       }
	    }
	 }
      };

      return req;
   }
}

/*---------------------------------------------------------------------*/
/*    jsonArgp ...                                                     */
/*    -------------------------------------------------------------    */
/*    Can the argument being serialized using the JSON format?         */
/*---------------------------------------------------------------------*/
function jsonArgp(n) {
   const t = typeof(n);
   return t === "number" || t === "string" || t === "boolean";
}

/*---------------------------------------------------------------------*/
/*    decode ...                                                       */
/*---------------------------------------------------------------------*/
function decode(req) {
   switch (req.getResponseHeader("Hop-Serialize")) {
      case "string": return req.response;
      case "number": return Number(req.response);
      case "json": return JSON.parse(req.response);
      case "null": return null;
      case "undefined": return undefined;
      case "true": return true;
      case "false": return false;
      default: return req.response;
   }
}

/*---------------------------------------------------------------------*/
/*    Service ...                                                      */
/*---------------------------------------------------------------------*/
function Service(base) {
   var svc = function() { return new Frame(this, base, arguments) };
   svc.base = base;
   svc.__proto__ = Service.prototype;
   return svc;
}

Service.prototype = {
   __proto__: Function.prototype
};

/*---------------------------------------------------------------------*/
/*    Server ...                                                       */
/*---------------------------------------------------------------------*/
class ClientServer {
   import(base) {
      return new Service(base);
   }
}

/*---------------------------------------------------------------------*/
/*    server ...                                                       */
/*---------------------------------------------------------------------*/
const server = new ClientServer();

/*---------------------------------------------------------------------*/
/*    hopHTMLTag ...                                                   */
/*---------------------------------------------------------------------*/
function hopHTMLTag(tag) {
   return (attrs, ...nodes) => {
      const el = document.createElement(tag);
      nodes.forEach(n => el.appendChild(n));
      return el;
   }
}

/*---------------------------------------------------------------------*/
/*    HTML tags ...                                                    */
/*---------------------------------------------------------------------*/
export const DIV = hopHTMLTag("DIV");
export const SPAN = hopHTMLTag("SPAN");
