// hop/node/hop.mjs
/*=====================================================================*/
/*    serrano/prgm/project/hop/3.7.x/node_modules/hop/node/hop.mjs     */
/*    -------------------------------------------------------------    */
/*    Author      :  manuel serrano                                    */
/*    Creation    :  Tue Oct 24 09:50:25 2023                          */
/*    Last change :  Fri Jan 19 15:56:44 2024 (serrano)                */
/*    Copyright   :  2023-24 manuel serrano                            */
/*    -------------------------------------------------------------    */
/*    ESM module                                                       */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Import/Export                                                    */
/*---------------------------------------------------------------------*/
import { createRequire } from 'node:module';
export * from "./html.mjs";

/*---------------------------------------------------------------------*/
/*    Dynamique require                                                */
/*---------------------------------------------------------------------*/
const require = createRequire(import.meta.url);

/*---------------------------------------------------------------------*/
/*    Default parameters                                               */
/*---------------------------------------------------------------------*/
export const isServer = true;
export const engine = "node";

const urlRoot = "/hop";
const port = 8888;

let express = false;

/*---------------------------------------------------------------------*/
/*    expressInit ...                                                  */
/*---------------------------------------------------------------------*/
function expressInit() {
   if (!express) {
      // this must be changed for ../../express
      express = require("../../../express")();
      express.listen(port);
   }
}

/*---------------------------------------------------------------------*/
/*    Service ...                                                      */
/*---------------------------------------------------------------------*/
/* @sealed */
export class Service {
   #fun;
   #url;
   
   constructor(fun, url) {
      this.#fun = fun;
      this.#url = url;
      
      if (!express) {
	 expressInit();
      }
   
      express.get(urlRoot + "/" + url, (req, res) => {
	 hopResponse(req, res, fun.call(req, req.query));
      });
   }

}

/*---------------------------------------------------------------------*/
/*    hopResponse ...                                                  */
/*---------------------------------------------------------------------*/
function hopResponse(req, res, val) {

   // keep-alive
   if (req.get("connection")) {
      res.set("Connection", req.Connection);
   }
   
   if (val instanceof response) {
      val.hopResponse(req, res);
   } else if (typeof(val) === "string") {
      res.send(val);
   } else if (val instanceof Promise) {
      val.then(val => hopResponse(req, res, val));
   } else {
      console.error("S=", val.toString());
      res.send(val.toString());
   }
}

/*---------------------------------------------------------------------*/
/*    response                                                         */
/*---------------------------------------------------------------------*/
export class response {
   val;
   constructor(val) {
      this.val = val;
   }
   hopResponse(req, res) {
      throw "not implemented";
   }
}

/*---------------------------------------------------------------------*/
/*    responseString ...                                               */
/*---------------------------------------------------------------------*/
export class responseString extends response {
   hopResponse(req, res) {
      res.send(this.val);
   }
}

/*---------------------------------------------------------------------*/
/*    responseFile ...                                                 */
/*---------------------------------------------------------------------*/
export class responseFile extends response {
   hopResponse(res) {
      res.sendFile(this.val);
   }
}
