// hop/node/html.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/3.7.x/node_modules/hop/node/html.mjs        */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Fri Jan 19 15:55:58 2024                          */
/*    Last change :  Fri Jan 19 16:25:50 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    Default HTML tags                                                */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    HtmlElement ...                                                  */
/*---------------------------------------------------------------------*/
/* @sealed */
class HtmlElement {
   #tag;
   #attrs;
   #nodes;
   #tostring  = false;

   constructor(tag, attrs, nodes) {
      this.#tag = tag;
      this.#attrs = attrs;
      this.#nodes = nodes;
   }

   static attrsToString(attrs) {
      return Object.keys(attrs)
	 .map(k => `${k}="${attrs[k].toString()"`).join(" ")
	 + ">";
   }
   
   toString() {
      if (!this.#tostring) {
	 this.#tostring = "<" + this.#tag
	    + (#this.attrs ? attrsToString(#this.attrs) : ">")
	    + this.#nodes.flatMap(n => n.toString()).join("")
	    + "</" + this.#tag + ">";
      }
      return this.#tostring;
   }
}

/*---------------------------------------------------------------------*/
/*    HTML tags ...                                                    */
/*---------------------------------------------------------------------*/
export function HTML(attrs, ...nodes) {
   return new HtmlElement("html", attrs, nodes);
}

export function HEAD(attrs, ...nodes) {
   return new HtmlElement("head", attrs, nodes);
}

export function SCRIPT(attrs, ...nodes) {
   return new HtmlElement("script", attrs, nodes);
}

export function STYLE(attrs, ...nodes) {
   return new HtmlElement("style", attrs, nodes);
}
