// hop/node/html.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/3.7.x/node_modules/hop/node/html.mjs        */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Fri Jan 19 15:55:58 2024                          */
/*    Last change :  Mon Jan 22 07:20:58 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    Default HTML tags                                                */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    HtmlElement ...                                                  */
/*---------------------------------------------------------------------*/
/* @sealed */
class HtmlElement {
   tag;
   attrs;
   nodes;
   tostring  = false;

   constructor(tag, attrs, nodes) {
      this.tag = tag;
      this.attrs = attrs;
      this.nodes = nodes;
   }

   static attrsToString(attrs) {
      return Object.keys(attrs)
	 .map(k => `${k}="${attrs[k].toString()}"`).join(" ")
	 + ">";
   }
   
   toString() {
      if (!this.tostring) {
	 this.tostring = "<" + this.tag + " "
	    + (this.attrs ? HtmlElement.attrsToString(this.attrs) : ">")
	    + this.nodes.flatMap(n => n.toString()).join("")
	    + "</" + this.tag + ">";
      }
      return this.tostring;
   }
}

/*---------------------------------------------------------------------*/
/*    Tilde                                                            */
/*---------------------------------------------------------------------*/
class Tilde extends HtmlElement {
   toString() {
      if (!this.tostring) {
	 this.tostring = this.nodes.flatMap(n => n.toString()).join("");
      }
      return this.tostring;
   }
}

/*---------------------------------------------------------------------*/
/*    HTML tags ...                                                    */
/*---------------------------------------------------------------------*/
export function BODY(attrs, ...nodes) {
   return new HtmlElement("body", attrs, nodes);
}

export function DIV(attrs, ...nodes) {
   return new HtmlElement("div", attrs, nodes);
}

export function HEAD(attrs, ...nodes) {
   return new HtmlElement("head", attrs, nodes);
}

export function HTML(attrs, ...nodes) {
   return new HtmlElement("html", attrs, nodes);
}

export function SCRIPT(attrs, ...nodes) {
   return new HtmlElement("script", attrs, nodes);
}

export function SPAN(attrs, ...nodes) {
   return new HtmlElement("span", attrs, nodes);
}

export function STYLE(attrs, ...nodes) {
   return new HtmlElement("style", attrs, nodes);
}

/*---------------------------------------------------------------------*/
/*    Pseudo HTML tags                                                 */
/*---------------------------------------------------------------------*/
export function TILDE(attrs, ...nodes) {
   return new Tilde("tilde", attrs, nodes);
}

export function DOLLAR(expr) {
   switch (typeof expr) {
      case "string": return "'" + expr + "'";
      case "number": return expr;
      default: return "";
   }
}
