// hop/node/html.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/3.7.x/node_modules/hop/node/html.mjs        */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Fri Jan 19 15:55:58 2024                          */
/*    Last change :  Tue Jan 30 16:12:11 2024 (serrano)                */
/*    Copyright   :  2024 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    HTML implementation                                              */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    XmlElement ...                                                   */
/*---------------------------------------------------------------------*/
/* @sealed */
class XmlElement {
   tag;
   attrs;
   nodes;
   tostring  = false;

   constructor(tag, attrs, nodes) {
      this.tag = tag;
      this.attrs = attrs;
      this.nodes = nodes;
   }

   static attrsToString(attrs) {
      return Object.keys(attrs)
	 .map(k => `${k}="${attrs[k].toString()}"`).join(" ");
   }
   
   toString() {
      if (!this.tostring) {
	 this.tostring = "<" + this.tag
	    + (this.attrs ? " " + XmlElement.attrsToString(this.attrs) + ">" : ">")
	    + this.nodes.flatMap(n => n.toString()).join("")
	    + "</" + this.tag + ">";
      }
      return this.tostring;
   }

   toResponse() {
      return this.toString();
   }
}

/*---------------------------------------------------------------------*/
/*    Tilde ...                                                        */
/*---------------------------------------------------------------------*/
/* @sealed */
class Tilde extends XmlElement {
   toString() {
      if (!this.tostring) {
	 this.tostring = this.nodes.flatMap(n => n.toString()).join("");
      }
      return this.tostring;
   }
}

/*---------------------------------------------------------------------*/
/*    HtmlElement ...                                                  */
/*---------------------------------------------------------------------*/
/* @sealed */
class HtmlElement extends XmlElement {
   mimetype = "text/html";
}

/*---------------------------------------------------------------------*/
/*    HTML tags ...                                                    */
/*---------------------------------------------------------------------*/
export function BODY(attrs, ...nodes) {
   return new XmlElement("body", attrs, nodes);
}

export function BUTTON(attrs, ...nodes) {
   return new XmlElement("button", attrs, nodes);
}

export function DIV(attrs, ...nodes) {
   return new XmlElement("div", attrs, nodes);
}

export function HEAD(attrs, ...nodes) {
   return new XmlElement("head", attrs, nodes);
}

export function HTML(attrs, ...nodes) {
   return new HtmlElement("html", attrs, nodes);
}

export function SCRIPT(attrs, ...nodes) {
   return new XmlElement("script", attrs, nodes);
}

export function SPAN(attrs, ...nodes) {
   return new XmlElement("span", attrs, nodes);
}

export function STYLE(attrs, ...nodes) {
   return new XmlElement("style", attrs, nodes);
}

export function TABLE(attrs, ...nodes) {
   return new XmlElement("table", attrs, nodes);
}

export function TD(attrs, ...nodes) {
   return new XmlElement("td", attrs, nodes);
}

export function TH(attrs, ...nodes) {
   return new XmlElement("th", attrs, nodes);
}

export function TR(attrs, ...nodes) {
   return new XmlElement("tR", attrs, nodes);
}

/*---------------------------------------------------------------------*/
/*    TILDE ...                                                        */
/*---------------------------------------------------------------------*/
export function TILDE(attrs, ...nodes) {
   return new Tilde("tilde", attrs, nodes);
}

/*---------------------------------------------------------------------*/
/*    DOLLAR ...                                                       */
/*---------------------------------------------------------------------*/
export function DOLLAR(expr) {
   switch (typeof expr) {
      case "string": return "'" + expr + "'";
      case "number": return expr;
      default: {
	 if ("dollar" in expr) {
	    return expr.dollar();
	 } else {
	    return `/* ${expr.toString()} */`;
	 }
      }
   }
}
