// syslog/hop/syslog.mjs
/*=====================================================================*/
/*    .../prgm/project/hop/hop/node_modules/syslog/hop/syslog.mjs      */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Jul  6 06:11:11 2014                          */
/*    Last change :  Wed May 29 08:25:44 2024 (serrano)                */
/*    Copyright   :  2014-24 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HOP syslog                                                       */
/*=====================================================================*/
"use hopscript"

/*---------------------------------------------------------------------*/
/*    ES6 exports                                                      */
/*---------------------------------------------------------------------*/
export { open, log, close } from "./_syslog.hop";
export { LOG_CONS, LOG_NDELAY, LOG_ODELAY, LOG_PID } from "./_syslog.hop";
export { LOG_AUTH, LOG_AUTHPRIV, LOG_CRON, LOG_DAEMON,
       	 LOG_FTP, LOG_KERN, 
	 LOG_LOCAL0, LOG_LOCAL1, LOG_LOCAL2, LOG_LOCAL3, LOG_LOCAL4, 
	 LOG_LOCAL5, LOG_LOCAL6, LOG_LOCAL7,
	 LOG_LPR, LOG_MAIL, LOG_NEWS, LOG_SYSLOG, LOG_USER, LOG_UUCP } from "./_syslog.hop";
export { LOG_EMERG, LOG_ALERT, LOG_CRIT, LOG_ERR,
       	 LOG_WARNING, LOG_NOTICE, LOG_INFO, LOG_DEBUG } from "./_syslog.hop";
