// spage/lib/spage.js
/*=====================================================================*/
/*    .../prgm/project/hop/3.0.x/node_modules/spage/lib/spage.js       */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Jul  6 06:11:11 2014                          */
/*    Last change :  Tue Sep 15 08:01:39 2015 (serrano)                */
/*    Copyright   :  2014-20 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HOP spage HTML widget                                            */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Bigloo module clause to spawn hopjs SPAGE initialization         */
/*---------------------------------------------------------------------*/
var hop = require( "hop" );
var config = require( hop.config );
var path = require( "path" );
var spage = require( "./_spage.hop" );

exports.SPAGE = spage.SPAGE;
exports.SPTAB = spage.SPTAB;
exports.SPHEAD = spage.SPHEAD;
exports.SPTAB = spage.SPTAB;
exports.SPTABHEAD = spage.SPTABHEAD;

exports.css = path.join( config.shareDir, "hop-spage.hss" );
exports.script = path.join( config.shareDir, "hopjs-spage.js" );
exports.jscript = exports.script;
