// spage/lib/spage.js
/*=====================================================================*/
/*    serrano/prgm/project/hop/hop/node_modules/spage/lib/spage.js     */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Jul  6 06:11:11 2014                          */
/*    Last change :  Fri Dec 31 17:31:21 2021 (serrano)                */
/*    Copyright   :  2014-21 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HOP spage HTML widget                                            */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Bigloo module clause to spawn hopjs SPAGE initialization         */
/*---------------------------------------------------------------------*/
var hop = require( "hop" );
var config = require( hop.config );
var path = require( "path" );
var spage = require( "./_spage.hop" );

const SPAGE = spage.SPAGE;
const SPTAB = spage.SPTAB;
const SPHEAD = spage.SPHEAD;
const SPTABHEAD = spage.SPTABHEAD;

const css = path.join( config.shareDir, "hop-spage.hss" );
const script = path.join( config.shareDir, "hopjs-spage.js" );

/*---------------------------------------------------------------------*/
/*    nodejs module                                                    */
/*---------------------------------------------------------------------*/
exports.SPAGE = SPAGE;
exports.SPTAB = SPTAB;
exports.SPHEAD = SPHEAD;
exports.SPTABHEAD = SPTABHEAD;

exports.css = css;
exports.script = script;
exports.jscript = exports.script;

/*---------------------------------------------------------------------*/
/*    es6 module                                                       */
/*---------------------------------------------------------------------*/
export { SPAGE, SPTAB, SPHEAD, SPTABHEAD, css, script, script as jscript };
