// fontifier/lib/fontifier.js
/*=====================================================================*/
/*    .../hop/3.0.x/node_modules/fontifier/lib/fontifier.js            */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Tue Dec 16 18:30:44 2014                          */
/*    Last change :  Sun Dec 21 11:37:27 2014 (serrano)                */
/*    Copyright   :  2014 Manuel Serrano                               */
/*    -------------------------------------------------------------    */
/*    HOP fontifier                                                    */
/*=====================================================================*/

/*---------------------------------------------------------------------*/
/*    Bigloo module clause to spawn hopjs FONTIFIER initialization     */
/*---------------------------------------------------------------------*/
var hop = require( "hop" );
var path = require( "path" );
var fontifier = require( "./_fontifier.hop" );

exports.javascriptFontifier = fontifier.jsFontifier;
exports.hopscriptFontifier = fontifier.hopscriptFontifier;
exports.hssFontifier = fontifier.hssFontifier;
exports.hopFontifier = fontifier.hopFontifier;
exports.lineNumber = fontifier.lineNumber;

exports.css = path.join( path.dirname( __dirname ), "hss/fontifier.hss" );
