// hz/lib/hz.js
/*=====================================================================*/
/*    serrano/prgm/project/hop/hop/node_modules/hz/lib/hz.js           */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Tue Mar 13 07:37:34 2018                          */
/*    Last change :  Sat Mar  5 19:56:28 2022 (serrano)                */
/*    Copyright   :  2018-22 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    HZ interface creation                                            */
/*=====================================================================*/
"use hopscript";

/*---------------------------------------------------------------------*/
/*    exports                                                          */
/*---------------------------------------------------------------------*/
export { install, installSync };
export { uninstall, uninstallSync };
export { remove, removeSync };
export { exists, existsSync };

/*---------------------------------------------------------------------*/
/*    web services                                                     */
/*---------------------------------------------------------------------*/
const http = hop.ports.https ? "https" : "http";
const localhost = `${http}://localhost:${hop.port || 8080}`;

/*---------------------------------------------------------------------*/
/*    install ...                                                      */
/*---------------------------------------------------------------------*/
function install(url, opt = undefined) {
   const host = opt?.host || localhost;
   const svc = hop.webService(host + "/hop/hz/install");
   return svc({url: url, checksum: opt?.checksum || false}).post();
}

/*---------------------------------------------------------------------*/
/*    installSync ...                                                  */
/*---------------------------------------------------------------------*/
function installSync(url, opt = undefined) {
   const host = opt?.host || "localhost";
   const svc = hop.webService(host + "/hop/hz/install");
   const o = svc({url: url, checksum: opt?.checksum || false}).postSync();
   console.log(o);
}

/*---------------------------------------------------------------------*/
/*    uninstall ...                                                    */
/*---------------------------------------------------------------------*/
function uninstall(name, opt = undefined) {
   const host = opt?.host || localhost;
   const svc = hop.webService(host + "/hop/hz/uninstall");
   return svc({name: name}).post();
}

/*---------------------------------------------------------------------*/
/*    uninstallSync ...                                                */
/*---------------------------------------------------------------------*/
function uninstallSync(name, opt = undefined) {
   const host = opt?.host || "localhost";
   const svc = hop.webService(host + "/hop/hz/uninstall");
   const o = svc({name: name}).postSync();
   console.log(o);
}

/*---------------------------------------------------------------------*/
/*    remove ...                                                       */
/*---------------------------------------------------------------------*/
function remove(name, opt = undefined) {
   const host = opt?.host || localhost;
   const svc = hop.webService(host + "/hop/hz/remove");
   return svc({name: name}).post();
}

/*---------------------------------------------------------------------*/
/*    removeSync ...                                                   */
/*---------------------------------------------------------------------*/
function removeSync(name, opt = undefined) {
   const host = opt?.host || "localhost";
   const svc = hop.webService(host + "/hop/hz/remove");
   const o = svc({name: name}).postSync();
   console.log(o);
}

/*---------------------------------------------------------------------*/
/*    exists ...                                                       */
/*---------------------------------------------------------------------*/
function exists(name, opt = undefined) {
   const host = opt?.host || localhost;
   const svc = hop.webService(host + "/hop/hz/exists");
   return svc({name: name}).post();
}

/*---------------------------------------------------------------------*/
/*    existsSync ...                                                   */
/*---------------------------------------------------------------------*/
function existsSync(name, opt = undefined) {
   const host = opt?.host || "localhost";
   const svc = hop.webService(host + "/hop/hz/exists");
   const o = svc({name: name}).postSync();
   console.log(o);
}
